rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py

```diff
@@ -0,0 +1,105 @@
+from rotkehlchen.chain.ethereum.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.chain.ethereum.structures import EthereumTxReceiptLog
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.constants.assets import A_ETH
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.types import EthereumTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from ..constants import CPT_UNISWAP_V2
+
+# https://www.4byte.directory/api/v1/event-signatures/?hex_signature=0xd78ad95fa46c994b6551d0da85fc275fe613ce37657fb8d5e3d130840159d822  # noqa: E501
+SWAP_SIGNATURE = b'\xd7\x8a\xd9_\xa4l\x99KeQ\xd0\xda\x85\xfc\'_\xe6\x13\xce7e\x7f\xb8\xd5\xe3\xd10\x84\x01Y\xd8"'  # noqa: E501
+
+
+class Uniswapv2Decoder(DecoderInterface):  # lgtm[py/missing-call-to-init]
+
+    def _maybe_decode_v2_swap(  # pylint: disable=no-self-use
+            self,
+            token: Optional[EvmToken],  # pylint: disable=unused-argument
+            tx_log: EthereumTxReceiptLog,
+            transaction: EthereumTransaction,  # pylint: disable=unused-argument
+            decoded_events: List[HistoryBaseEntry],
+            action_items: List[ActionItem],  # pylint: disable=unused-argument
+    ) -> None:
+        """Decode trade for uniswap v2. The approach is to read the events and detect the ones
+        where the user sends and receives any asset. The spend asset is the swap executed and
+        the received is the result of calling the swap method in the uniswap contract.
+        We need to make sure that the events related to the swap are consecutive and for that
+        we make use of the maybe_reshuffle_events.
+        """
+        if tx_log.topics[0] == SWAP_SIGNATURE:
+            # When the router chains multiple swaps in one transaction only the last swap has
+            # the buyer in the topic. In that case we know it is the last swap and the receiver is
+            # the user
+            maybe_buyer = hex_or_bytes_to_address(tx_log.topics[2])
+            out_event = in_event = None
+
+            # amount_in is the amount that enters the pool and amount_out the one
+            # that leaves the pool
+            amount_in_0, amount_in_1 = hex_or_bytes_to_int(tx_log.data[0:32]), hex_or_bytes_to_int(tx_log.data[32:64])  # noqa: E501
+            amount_out_0, amount_out_1 = hex_or_bytes_to_int(tx_log.data[64:96]), hex_or_bytes_to_int(tx_log.data[96:128])  # noqa: E501
+            amount_in, amount_out = amount_in_0, amount_out_0
+            if amount_in == ZERO:
+                amount_in = amount_in_1
+            if amount_out == ZERO:
+                amount_out = amount_out_1
+
+            received_eth = ZERO
+            for event in decoded_events:
+                if event.asset == A_ETH and event.event_type == HistoryEventType.RECEIVE:
+                    received_eth += event.balance.amount
```

Comment:
Also where does it come from?

Reply:
Where does what come from? The ETH? It's the `value` part of the transaction

Reply:
This is because when trading ETH for tokens you have to send all the ETH you want to trade but the price could change and you end up trading less ETH than what you wanted. So we check the amount of ETH received in the transaction. When decoding the SPEND part of the transaction you have to verify that the amount traded + the amount returned to the user match the one you send